File: backend-node/routes/creators.js
const express = require('express');
const router = express.Router();
const { pool } = require('../db');
const { authMiddleware, crypto } = require('../auth');

async function getCreator(req) {
  const { rows } = await pool.query('SELECT * FROM creators WHERE user_id = $1 AND status = $2', [req.userId, 'active']);
  return rows[0];
}

router.post('/register', authMiddleware, async (req, res, next) => {
  try {
    const { pen_name, real_name, introduction, contact_email, contact_phone } = req.body;
    if (!pen_name) return res.status(400).json({ message: '请输入笔名' });
    
    const existing = await getCreator(req);
    if (existing) return res.status(400).json({ message: '您已是创作者' });
    
    const id = crypto.randomUUID();
    const social = {};
    if (contact_phone) social.phone = contact_phone;
    await pool.query(
      'INSERT INTO creators (id, user_id, pen_name, real_name, introduction, contact_email, contact_phone, social_accounts) VALUES ($1, $2, $3, $4, $5, $6, $7, $8)',
      [id, req.userId, pen_name, real_name || null, introduction || '', contact_email || null, contact_phone || null, JSON.stringify(social)]
    );
    await pool.query('UPDATE users SET is_creator = TRUE WHERE id = $1', [req.userId]);
    const creator = (await pool.query('SELECT * FROM creators WHERE id = $1', [id])).rows[0];
    res.status(201).json(creator);
  } catch (e) { next(e); }
});

router.get('/profile', authMiddleware, async (req, res, next) => {
  try {
    const creator = await getCreator(req);
    if (!creator) return res.status(403).json({ message: '还不是创作者' });
    res.json(creator);
  } catch (e) { next(e); }
});

router.get('/novels', authMiddleware, async (req, res, next) => {
  try {
    const creator = await getCreator(req);
    if (!creator) return res.status(403).json({ message: '还不是创作者' });
    const page = Math.max(1, parseInt(req.query.page) || 1);
    const pageSize = 20;
    const { rows } = await pool.query(
      'SELECT id, title, author, cover_url, description, status, word_count, chapter_count, view_count, rating FROM novels WHERE creator_id = $1 ORDER BY created_at DESC LIMIT $2 OFFSET $3',
      [creator.id, pageSize, (page - 1) * pageSize]
    );
    const countRow = await pool.query('SELECT COUNT(*) as count FROM novels WHERE creator_id = $1', [creator.id]);
    res.json({ total: parseInt(countRow.rows[0].count), page, page_size: pageSize, items: rows });
  } catch (e) { next(e); }
});

router.get('/dashboard', authMiddleware, async (req, res, next) => {
  try {
    const creator = await getCreator(req);
    if (!creator) return res.status(403).json({ message: '还不是创作者' });
    const stats = await pool.query(
      'SELECT COUNT(*) as total_novels, COALESCE(SUM(word_count), 0) as total_words, COALESCE(SUM(view_count), 0) as total_views, COALESCE(SUM(like_count), 0) as total_likes, COALESCE(SUM(collect_count), 0) as total_collects, COALESCE(SUM(chapter_count), 0) as total_chapters FROM novels WHERE creator_id = $1',
      [creator.id]
    );
    const recent = await pool.query('SELECT id, title, author, word_count, chapter_count, view_count FROM novels WHERE creator_id = $1 ORDER BY created_at DESC LIMIT 5', [creator.id]);
    await pool.query('UPDATE creators SET total_words = $1, total_novels = $2, total_readers = $3 WHERE id = $4',
      [stats.rows[0].total_words, stats.rows[0].total_novels, stats.rows[0].total_views, creator.id]);
    res.json({ ...stats.rows[0], recent_novels: recent.rows });
  } catch (e) { next(e); }
});

module.exports = router;

File: backend/routers/creators.py
"""
创作者路由: 注册创作者 / 创作者档案 / 创作者作品列表 / 仪表盘数据
"""
import math
from typing import List
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from sqlalchemy import func

from database import get_db
from models import Creator, User, Novel
from schemas import CreatorRegister, CreatorOut, NovelBrief, MessageResponse
from security import get_current_user, get_current_creator

router = APIRouter(prefix="/api/creator", tags=["创作者"])


@router.post("/register", response_model=CreatorOut, status_code=201)
def register_creator(
    data: CreatorRegister,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    if user.is_creator and user.creator_profile:
        raise HTTPException(status_code=400, detail="您已经是创作者")

    creator = Creator(
        user_id=user.id,
        pen_name=data.pen_name,
        real_name=data.real_name,
        introduction=data.introduction or "",
        contact_email=data.contact_email,
        contact_phone=data.contact_phone,
        social_accounts=data.social_accounts or {},
    )
    db.add(creator)
    user.is_creator = True
    if not user.nickname:
        user.nickname = data.pen_name
    db.commit()
    db.refresh(creator)
    return creator


@router.get("/profile", response_model=CreatorOut)
def my_profile(creator=Depends(get_current_creator)):
    return creator


@router.put("/profile", response_model=CreatorOut)
def update_profile(
    data: dict,
    creator=Depends(get_current_creator),
    db: Session = Depends(get_db),
):
    allowed = ["pen_name", "introduction", "contact_email", "contact_phone", "social_accounts"]
    for k, v in data.items():
        if k in allowed:
            setattr(creator, k, v)
    db.commit()
    db.refresh(creator)
    return creator


@router.get("/novels", response_model=dict)
def my_novels(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    creator=Depends(get_current_creator),
    db: Session = Depends(get_db),
):
    query = db.query(Novel).filter(Novel.creator_id == creator.id).order_by(Novel.created_at.desc())
    total = query.count()
    items = query.offset((page - 1) * page_size).limit(page_size).all()
    return {
        "total": total,
        "page": page,
        "page_size": page_size,
        "items": [NovelBrief.model_validate(n).model_dump() for n in items],
    }


@router.get("/dashboard", response_model=dict)
def dashboard(creator=Depends(get_current_creator), db: Session = Depends(get_db)):
    """创作者仪表盘统计数据"""
    novels = db.query(Novel).filter(Novel.creator_id == creator.id).all()
    total_novels = len(novels)
    total_words = sum(n.word_count for n in novels)
    total_views = sum(n.view_count for n in novels)
    total_likes = sum(n.like_count for n in novels)
    total_collects = sum(n.collect_count for n in novels)
    total_chapters = sum(n.chapter_count for n in novels)

    # 更新创作者统计
    creator.total_words = total_words
    creator.total_novels = total_novels
    creator.total_readers = total_views
    db.commit()

    return {
        "total_novels": total_novels,
        "total_words": total_words,
        "total_views": total_views,
        "total_likes": total_likes,
        "total_collects": total_collects,
        "total_chapters": total_chapters,
        "recent_novels": [NovelBrief.model_validate(n).model_dump() for n in novels[:5]],
    }


@router.get("/list", response_model=List[CreatorOut])
def list_creators(
    limit: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
):
    """公开的创作者列表"""
    creators = db.query(Creator).filter(
        Creator.status == "active"
    ).order_by(Creator.total_readers.desc()).limit(limit).all()
    return creators
